src/host-config.ts

```diff
@@ -0,0 +1,174 @@
+const noProxyPath = path.join(proxyBasePath, 'no_proxy');
+
+interface ProxyConfig {
+	[key: string]: string | string[];
+}
+
+interface HostConfig {
+	network: {
+		proxy?: ProxyConfig | null;
+		hostname?: string | null;
+	};
+}
+
+const isAuthField = (field: string): boolean =>
+	_.includes(['login', 'password'], field);
+
+async function readProxy(): Promise<ProxyConfig | null> {
+	const conf: ProxyConfig = {};
+	try {
+		const redsocksConf = await fs.readFile(redsocksConfPath, 'utf-8');
+		const lines = redsocksConf.split('\n');
+
+		for (const line of lines) {
+			for (const proxyField of proxyFields) {
+				let match: string[] | null = null;
+				if (isAuthField(proxyField)) {
+					match = line.match(new RegExp(proxyField + '\\s*=\\s*"(.*)"\\s*;'));
+				} else {
+					match = line.match(
+						new RegExp(proxyField + '\\s*=\\s*([^;\\s]*)\\s*;'),
+					);
+				}
+
+				if (match != null) {
+					conf[proxyField] = match[1];
+				}
+			}
+		}
+	} catch (e) {
+		if (!ENOENT(e)) {
+			throw e;
+		}
+		return null;
+	}
+
+	try {
+		const noProxy = await fs.readFile(noProxyPath, 'utf-8');
+		conf.noProxy = noProxy.split('\n');
+	} catch (e) {
+		if (!ENOENT(e)) {
+			throw e;
+		}
+	}
+	return conf;
+}
+
+function generateRedsocksConfEntries(conf: ProxyConfig): string {
+	let val = '';
+	for (const field of proxyFields) {
+		if (conf[field] != null) {
+			let v = conf[field];
+			if (isAuthField(field)) {
+				v = `"${v}"`;
+			}
+			val += `\t${field} = ${v};\n`;
+		}
+	}
+	return val;
+}
+
+async function setProxy(conf: ProxyConfig | null): Promise<void> {
+	if (_.isEmpty(conf) || conf == null) {
+		try {
+			await Bluebird.all([fs.unlink(redsocksConfPath), fs.unlink(noProxyPath)]);
+		} catch (e) {
+			if (!ENOENT(e)) {
+				throw e;
+			}
+		}
+	} else {
+		await mkdirp(proxyBasePath);
+		if (_.isArray(conf.noProxy)) {
+			await writeFileAtomic(noProxyPath, conf.noProxy.join('\n'));
+		}
+		const redsocksConf = `${redsocksHeader}${generateRedsocksConfEntries(
+			conf,
+		)}${redsocksFooter}`;
+		await writeFileAtomic(redsocksConfPath, redsocksConf);
+	}
+
+	await systemd.restartService('resin-proxy-config');
+	await systemd.restartService('redsocks');
+}
+
+const hostnamePath = path.join(constants.rootMountPoint, '/etc/hostname');
+async function readHostname() {
+	const hostnameData = await fs.readFile(hostnamePath, 'utf-8');
+	return _.trim(hostnameData);
+}
+
+async function setHostname(val: string, configModel: Config) {
+	await configModel.set({ hostname: val });
+	await systemd.restartService('resin-hostname');
+}
+
+// Don't use async/await here to maintain the bluebird
+// promises being returned
+export function get(): Bluebird<HostConfig> {
+	return Bluebird.join(readProxy(), readHostname(), (proxy, hostname) => {
+		return {
+			network: {
+				proxy,
+				hostname,
+			},
+		};
+	});
+}
+
+export async function patch(
+	conf: HostConfig,
+	configModel: Config,
+): Bluebird<void> {
```

Comment:
If you change this to `Promise<void>` you can drop the `return Bluebird.resolve()` below

Reply:
I'm trying to maintain the external interface for now, no doubt there's somewhere in CS where it's assumed to be a Bluebird promise. Once it's all typescript, I can change the types and get the compiler to find those places for me, but it's a right pain before.